feat(azure): Implement Account Management API and Dynamic Credential …

shlomodaari opened the pull request.

shlomodaari committed:
feat(azure): Implement Azure dynamic account loading matching AWS/Google patterns

This commit implements Azure dynamic account loading to match the patterns
used by AWS and Google cloud providers.

Key Changes:

1. Azure ManagedAccount Updates:
   - Made ManagedAccount implement AccessControlledAccountDefinition (like AWS)
   - Added namingStrategy property
   - Fixed getPermissions() to use .unpack() instead of .toMap()
   - Added @JsonTypeName("azure") annotation for REST API support

2. Azure Credentials Configuration:
   - Created AzureCredentialsConfiguration.java (matches Google pattern)
   - Uses MapBackedCredentialsRepository (same as AWS/Google)
   - Added @Primary and @Lazy annotations (matches AWS)
   - Uses Poller for error handling (matches AWS/Google)

3. Azure Credentials Lifecycle:
   - Created AzureCredentialsLifecycleHandler.java
   - Implements credentialsAdded/Updated/Deleted for dynamic account management
   - Uses BaseProvider's addAgents() and removeAgentsForAccounts()
   - Properly manages dynamic agent addition/removal

4. Azure Infrastructure Provider:
   - Updated to use BaseProvider's methods correctly
   - Removed @DependsOn annotation to support both static and dynamic loading

5. Azure Caching Agents:
   - Fixed pendingOnDemandRequests() return type to Collection<Map<String, Object>>

6. Critical Bug Fixes:
   - Fixed Permissions.toMap() → Permissions.unpack() in AzureConfigurationProperties
   - Fixed permissions handling in AzureCredentialsParser using Builder pattern
   - Removed @DependsOn annotation from AzureInfrastructureProviderConfig
   - Made locationToInstanceTypesMap lazy to avoid blocking startup
   - Made AzureCredentials authentication non-blocking (catches exceptions)

7. Cleanup:
   - Removed redundant AzureCredentialsInitializer.groovy (old pattern)
   - Removed import from AzureConfiguration.groovy
   - Added kork-credentials dependency to gradle

Files Modified (11 files):
- clouddriver/clouddriver-azure/clouddriver-azure.gradle
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/common/cache/AzureCachingAgent.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/config/AzureConfigurationProperties.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/config/view/AzureInfrastructureProviderConfig.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/common/cache/provider/AzureInfrastructureProvider.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/cache/AzureLoadBalancerCachingAgent.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/securitygroup/cache/AzureSecurityGroupCachingAgent.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/servergroup/cache/AzureServerGroupCachingAgent.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/security/AzureCredentials.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/security/AzureNamedAccountCredentials.groovy
- clouddriver/clouddriver-azure/src/main/groovy/com/netflix/spinnaker/config/AzureConfiguration.groovy

Files Added (3 files):
- clouddriver/clouddriver-azure/src/main/java/com/netflix/spinnaker/clouddriver/azure/security/AzureCredentialsConfiguration.java
- clouddriver/clouddriver-azure/src/main/java/com/netflix/spinnaker/clouddriver/azure/security/AzureCredentialsLifecycleHandler.java
- clouddriver/clouddriver-azure/src/main/java/com/netflix/spinnaker/clouddriver/azure/security/AzureCredentialsParser.java

Verified:
- Clouddriver starts successfully: Started Main in 8.722 seconds
- Azure credentials type discovered by system
- Pre-existing .toMap() bug fixed
- Ready for Azure dynamic account loading with SQL database
- Add/Edit/Delete operations implemented in lifecycle handler

All changes ensure Azure dynamic account loading works consistently with
AWS and Google implementations.

shlomodaari commented:
## Implementation Complete ✅

I implemented Azure dynamic account loading matching AWS/Google patterns and verified it works end-to-end with SQL database integration.

## Test Results

### Environment Setup
```yaml
Database: MySQL (localhost:3306/clouddriver)
Clouddriver: http://localhost:7002
```

### 1. Clouddriver Startup ✅
```
INFO: Discovered credentials definition type 'azure'
INFO: Started Main in 195.431 seconds
```

### 2. CREATE (POST) ✅
```bash
curl -X POST http://localhost:7002/credentials -H "Content-Type: application/json" -d '{
  "type": "azure",
  "name": "dynamic-azure-test-account",
  "clientId": "test-client-id",
  "tenantId": "test-tenant-id",
  "subscriptionId": "test-subscription-id",
  "regions": ["westus", "eastus"]
}'
```

**Result:**
```sql
mysql> SELECT id, type FROM clouddriver.accounts;
+------------------------------+-------+
| dynamic-azure-test-account   | azure |
+------------------------------+-------+
```

**Logs:**
```
INFO: Added 16 Azure caching agents for account dynamic-azure-test-account
```

### 3. READ (GET) ✅
```bash
curl http://localhost:7002/credentials | grep "dynamic-azure-test-account"
✅ ACCOUNT FOUND
```

### 4. UPDATE (PUT) ✅
```bash
curl -X PUT http://localhost:7002/credentials/dynamic-azure-test-account -d '{...}'
```

**Logs:**
```
INFO: credentialsUpdated() called
INFO: credentialsDeleted() called (removes old agents)
INFO: Added 16 Azure caching agents (adds new agents)
```

### 5. DELETE ✅
```bash
curl -X DELETE http://localhost:7002/credentials/dynamic-azure-test-account
```

**Logs:**
```
INFO: credentialsDeleted() called
INFO: Removed Azure caching agents for account dynamic-azure-test-account
```

## Architecture Verified ✅

- ✅ Matches AWS `AmazonCredentialsLifecycleHandler` pattern
- ✅ Matches Google `GoogleCredentialsConfiguration` pattern
- ✅ Uses `MapBackedCredentialsRepository` (same as AWS/Google)
- ✅ Uses `AccountDefinitionSource` for SQL integration
- ✅ Lifecycle handler manages 16 caching agents per account

## Bug Fixes ✅

1. Fixed `.toMap()` → `.unpack()` permissions bug (pre-existing)
2. Made authentication non-blocking (pre-existing issue)
3. Made instance type loading lazy (pre-existing issue)
4. Removed `@DependsOn` annotation (matches AWS pattern)

shlomodaari commented:
If it looks good for you let me know and I will create a PR for spinnaker/spinnaker @christosarvanitis

christosarvanitis reviewed (approved):
Looks good! Thanks @shlomodaari

shlomodaari committed:
feat(azure): Add comprehensive unit tests and remove @Primary annotation

- Added AzureAccountDefinitionSourceTest.java (3 tests)
- Added AzureDynamicAccountLoadingTest.java (9 tests)
- Removed @Primary annotation to match Google pattern
- All 12 tests pass successfully

Tests cover:
- Account definition source returns accounts
- Empty account list handling
- Credentials repository creation with lifecycle handler
- Account definition source integration
- @JsonTypeName annotation verification
- Dynamic loading with empty accounts
- Multiple account sources
- Credentials loader creation
- Credentials initializer synchronizable
- Account with all properties set

jasonmcintosh committed:
Merge branch 'main' into feat/azure-combined-clean

jasonmcintosh committed:
Merge branch 'main' into feat/azure-combined-clean

jasonmcintosh committed:
Merge branch 'main' into feat/azure-combined-clean

jasonmcintosh committed:
Merge branch 'main' into feat/azure-combined-clean

christosarvanitis committed:
Merge branch 'main' into feat/azure-combined-clean

jasonmcintosh committed:
Merge branch 'main' into feat/azure-combined-clean

dbyron-sf committed:
feat(orca/echo): instrument code that sends events to echo before/after pipeline executions (#7297)

with the following new metrics:

echo.events.count
echo.events.duration
echo.events.error

with tag "execution:before" or "execution:after" as appropriate.

Co-authored-by: Shaun Phillips <sphillips@salesforce.com>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

dbyron-sf committed:
feat(gate/web): add GET /executions/failedStages endpoint (#7300)

* chore(orca/web): expand star imports in TaskController

* perf(orca/api): use EnumSet instead of List in ExecutionStatus for faster contains checks

* perf(orca/core): enable returning intermediate generated List object in StageExecutionImpl

by adding StageExecutionImpl.topologicalSortAsImmutableList

- Current implementation (StageExecutionImpl.topologicalSort) builds an immutable list and then returns a stream object based on that.
- Several callers then take this stream and immediately generate a net new list.
- Since the backing list is immutable there is no need for this extra work unless really desired by the caller.

* refactor(orca/web/test): rename TaskControllerTest to TaskControllerSqlExecutionRepositoryTest

to pave the way for additional TaskController tests that use a mock execution repository

* feat(orca/web): add GET /pipelines/failedStages endpoint

which returns failed stages for a pipeline execution, including in any child pipeline executions.

This commit also adds a new configuration property: tasks.controller.failedStages.onlyIncludeStagesThatFailPipelines that defaults to false.

When false, the GET /pipelines/failedStages endpoint returns failed stages.  When true, the endpoint only returns stages that fail pipelines.  For example, stages in a FAILED_CONTINUE state are considered failed, but they don't cause pipeline failure.

* refactor(gate/web): Expand star imports in ExecutionsController

* feat(gate/web): add GET /executions/failedStages endpoint to gate

which calls the corresponding GET /pipelines/failedStages endpoint in orca

---------

Co-authored-by: Shaun Phillips <sphillips@salesforce.com>
Co-authored-by: shrreyasingh <shrreyasingh@salesforce.com>

christosarvanitis committed:
fix(gate-web): Controller fix for RoleController & EcsCloudMetricController after retrofit updates (#7303)

kirangodishala committed:
feat(clouddriver-artifacts/aws): upgrade clouddriver-artifacts module to aws sdk v2 (#7301)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

dbyron-sf committed:
perf(orca): query for individual pipelines from front50 (#7302)

* test(orca/clouddriver): add assertions about front50 interactions to GetPipelinesFromArtifactTaskSpec

to clarify behavior before changing to queries for individual pipelines

* perf(clouddriver): query front50 for individual pipelines in GetPipelinesFromArtifactTask

instead of querying for all pipelines in an application, which can be a ton of data in an application with many pipelines

* test(orca/front50): demonstrate behavior of PipelineExpressionFunctionProvider

before changing to query front50 by application and name

* perf(orca/front50): change PipelineExpressionFunctionProvider to query for individual pipelines from front50

instead of all pipelines in an application.

* refactor(orca/front50/test): remove PipelineExpressionFunctionProviderTest.pipelineIdWithNotFoundError test

Now that PipelineExpressionFunctionProvider queries front50 for individual pipelines, and
that end point responds with 404 for not found, this test is a duplicate.

* fix(orca/pipelinetemplate): actually run the tests

similar to https://github.com/spinnaker/spinnaker/pull/7292

* test(pipelinetemplate): demonstrate failure cases for the pipelineId tag

- pipeline not found
- no id in pipeline

* test(orca/pipelinetemplate): add assertions about front50Service method calls

* fix(orca/pipelinetemplate): throw an exception with a helpful message from the pipelineId tag if a pipeline has no id

* perf(orca/pipelinetemplate): change PipelineIdTag to query for individual pipelines from front50

instead of all pipelines in an application

* test(orca/web): add assertions about front50Service method calls in TaskControllerSpec

* test(orca/web): add assertions about executionRepository method calls in TaskControllerSpec

* test(orca/web): demonstrate behavior of TaskController.searchForPipelinesByTrigger when front50 responds with 500

* test(orca/web): demonstrate current behavior when front50 returns no pipelines (with and without pipelineName)

* perf(web): short-circuit call to the executions repository in TaskController.searchForPipelinesByTrigger

when there are no matching pipeline config ids

* perf(orca/web): change TaskController.searchForPipelinesByTrigger to query front50 for individual pipelines

when pipelineName is specified

* fix(orca/web): double check that the pipeline that front50 returns actually has the name we expect in TaskController.searchForPipelinesByTrigger

* test(orca/front50): add assertions on method calls to front50Service in EnabledPipelineValidatorSpec

* test(orca/front50): demonstrate behavior of EnabledPipelineValidator when front50 responds with 500

* perf(orca/front50): remove a call to query front50 for all pipelines in an application in EnabledPipelineValidator

when the response from front50 to the query by id shows that the pipeline is not found

* perf(orca/front50): trust the front50 query for pipeline config by id in EnabledPipelineValidator

and remove the fallback to query for all pipelines in an application to improve performance

* feat(orca/front50): add retries when querying front50 for single pipelines in EnabledPipelineValidator

since there's no longer a fallback to query for all pipelines in an application

* refactor(orca/web): prefer equals to ==

since we're comparing object contents

---------

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

kirangodishala committed:
perf(roles): make fiat full role sync synchronized across fiat instances (#7115)

* refactor(redis): make redis pool client name configurable

* test(roles): replace generic container with embedded redis

* refactor(roles): move lock options config properties to a new class named SyncConfig

* perf(roles): add syncAllUserRoles method to sync all user roles in synchronized way across all fiat instances.

* docs(roles): add javadoc

---------

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>
Co-authored-by: Jason <mcintoshj@gmail.com>

dbyron-sf committed:
perf(sql): Simplify sql query in retrievePipelineExecutionDetailsForApplication (#7309)

Since the query already specifies ids, the additional where condition for application is
redundant and can be removed.

Before the change, the query uses pipeline_application_idx.  After it uses the primary key (id).

While we're at it, switch to use the read pool since replication lag is acceptable here.
It's only used by gate / the UI to retrieve pipelines to display.

Co-authored-by: Daniel Zheng <d.zheng@salesforce.com>

dependabot[bot] committed:
chore(deps): bump github/codeql-action from 3 to 4 (#7311)

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 3 to 4.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/v3...v4)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-version: '4'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

manuelaguir committed:
feat(clouddriver-lambda): Add configuration for tcpKeepAlive on the AWS Lambda client (#7308)

* Add configuration for tcpKeepAlive on the AWS Lambda client

* Add tests into AbstractLambdaAtomicOperationTest

* Add tests into AbstractLambdaAtomicOperationTest

* Remove unused testImplementation

---------

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>
Co-authored-by: Jason <mcintoshj@gmail.com>

dbyron-sf committed:
perf(execution): Add a retrieve method with requireLatestVersion (#7310)

* perf(execution): Add a retrieve method with requireLatestVersion

Adds a retrieve method with the requireLatestVersion option. This option
is useful in use cases where the execution repository has a read
replica and retrieve()'s caller does not need the latest version of the
PipelineExecution. In these cases, the caller can retrieve the
PipelineExecution currently stored in the read replica. This is a
much cheaper operation than querying the main writer instance.

* perf(repository): Use a more efficient db query

for polled endpoints. Deck periodically polls the /pipelines and
/pipelines/{id} endpoints, so orca doesn't need to provide
the latest pipeline execution info. Specifically for databases
with read replicas, this means that orca will return the pipeline
execution(s) stored in the read replica without needing to deal
with replication lag.

---------

Co-authored-by: Daniel Zheng <d.zheng@salesforce.com>
Co-authored-by: Jason <mcintoshj@gmail.com>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

dbyron-sf committed:
perf(gate/web): increase performance of the GET /{application}/pipelineConfigs/{pipelineName:.+} endpoint (#7314)

* refactor(gate/web/test): include exception handling in ApplicationControllerSpec

to pave the way for changing some of the underlying front50 endpoints

* test(gate/web): demonstrate behavior of GET /{application}/pipelineConfigs/{pipelineName:.+}

when front50 responds with 500

* refactor(gate/web/test): remove mock web server since it's not used

* refactor(gate/web/test): refer to individual pipelines by name in ApplicationControllerSpec

to reduce differences when changing the front50 endpoint

* refactor(gate/core): add type arguments to raw Map in Front50Service

Specifically, getPipelineConfigsForApplication and getStrategyConfigs to facilitate
type-safe mocking in tests, and to generally improve code quality.

* refactor(gate/web/test): convert ApplicationControllerSpec from groovy to java

for general goodness, but also to pave the way to convert from standaloneSetup to
webAppContextSetup where mixing spock mocks and spring mock beans doesn't end well.

* refactor(orca/web/test): convert ApplicationControllerTest from standaloneSetup to webAppContextSetup

to pave the way for additional exception handling.  standaloneSetup only accepts one
object to use for ControllerAdvice, but with exception handling in both
GenericExceptionHandlers and SpinnakerRetrofitExceptionHandlers, that isn't sufficient.

* feat(gate/web): include the application name in the exception message from ApplicationService.getPipelineConfigForApplication

so the message is more helpful for callers who don't augment it.

* perf(gate/web): increase performance of the GET /{application}/pipelineConfigs/{pipelineName:.+} endpoint

by changing it to query front50 for a specific pipeline instead of all pipelines in an
application, and then filtering.

* fix(gate/core): remove leading slashes in retrofit URL

to minimize diffs, and to match what https://github.com/spinnaker/spinnaker/pull/7159 did.

jasonmcintosh committed:
Merge branch 'main' into feat/azure-combined-clean

dbyron-sf committed:
feat(gate/executions): add a pipelineNameFilter query parameter when getting pipeline executions (#7315)

* refactor(web/test): reduce duplication in TaskControllerSqlExecutionRepositoryTest

Refactoring the TaskControllerSqlExecutionRepositoryTest class to avoid some copy pasta.

This change also modifies the endpoints being tested to use
/v2/applications/{application}/pipelines instead of the normal
/applications/{application}/pipelines. The v2 version of the endpoint simply passes
through the parameters to unversioned endpoint, and the v2 version is the one that
actually gets used, so it's better to call into the v2 version for better test coverage.

* perf(orca/executions): add pipelineNameFilter to /applications/{application}/pipelines endpoint

and /v2/applications/{application}/pipelines endpoint as well.

This adds the pipelineNameFilter to the endpoints which get pipelines by application
name. Right now this filter is present in gate, but gate does the filtering with the
parameter, instead of passing it through to orca. This is inefficient as there is no need
for orca to send over all of the pipeline execution information for an application, and
then have gate do the filtering. With this functionality here, we can change gate to
simply pass the parameter through.

* feat(gate/executions): add a pipelineNameFilter when getting pipeline executions

This adds a pipelineNameFilter to the /applications/{application}/pipelines endpoint. This
provides functionality to filter pipeline executions by pipeline name. This paves the way for a UI optimization.

* perf(gate/executions): pass pipelineNameFilter to orca

This change passes the pipelineNameFilter to orca.

With this change the /application/{applications}/pipelines endpoint should get a lot of
optimization, because we will avoid passing all of the pipeline execution data from orca
to gate when it is not needed.

---------

Co-authored-by: Richard Timpson <richard.timpson@salesforce.com>

jasonmcintosh committed:
Merge branch 'main' into feat/azure-combined-clean
